src/modules/ekf2/EKF/mag_fusion.cpp

```diff
@@ -230,98 +230,100 @@ bool Ekf::fuseMag(const Vector3f &mag, estimator_aid_source3d_s &aid_src_mag, bo
+		// check if the innovation variance calculation is badly conditioned
+		if (aid_src_status.innovation_variance >= aid_src_status.observation_variance) {
+			// the innovation variance contribution from the state covariances is not negative, no fault
+			_fault_status.flags.bad_hdg = false;
 
-		// we reinitialise the covariance matrix and abort this fusion step
-		initialiseCovariance();
-		ECL_ERR("yaw fusion numerical error - covariance reset");
-		return false;
-	}
+		} else {
+			// the innovation variance contribution from the state covariances is negative which means the covariance matrix is badly conditioned
+			_fault_status.flags.bad_hdg = true;
 
-	// calculate the Kalman gains
-	// only calculate gains for states we are using
-	Vector24f Kfusion;
+			// we reinitialise the covariance matrix and abort this fusion step
+			initialiseCovariance();
+			ECL_ERR("yaw fusion numerical error - covariance reset");
 
-	for (uint8_t row = 0; row < _k_num_states; row++) {
-		for (uint8_t col = 0; col <= 3; col++) {
-			Kfusion(row) += P(row, col) * H_YAW(col);
+			return false;
 		}
 
-		Kfusion(row) *= heading_innov_var_inv;
-	}
+		// calculate the Kalman gains
+		// only calculate gains for states we are using
+		Vector24f Kfusion;
+		const float heading_innov_var_inv = 1.f / aid_src_status.innovation_variance;
 
-	// define the innovation gate size
-	float gate_sigma = math::max(_params.heading_innov_gate, 1.f);
+		for (uint8_t row = 0; row < _k_num_states; row++) {
+			for (uint8_t col = 0; col <= 3; col++) {
+				Kfusion(row) += P(row, col) * H_YAW(col);
+			}
 
-	// innovation test ratio
-	setEstimatorAidStatusTestRatio(aid_src_status, gate_sigma);
+			Kfusion(row) *= heading_innov_var_inv;
+		}
 
-	// set the magnetometer unhealthy if the test fails
-	if (aid_src_status.innovation_rejected) {
-		_innov_check_fail_status.flags.reject_yaw = true;
-
-		// if we are in air we don't want to fuse the measurement
-		// we allow to use it when on the ground because the large innovation could be caused
-		// by interference or a large initial gyro bias
-		if (!_control_status.flags.in_air
-		&& isTimedOut(_time_last_in_air, (uint64_t)5e6)
-		&& isTimedOut(aid_src_status.time_last_fuse, (uint64_t)1e6)
-		) {
-			// constrain the innovation to the maximum set by the gate
-			// we need to delay this forced fusion to avoid starting it
-			// immediately after touchdown, when the drone is still armed
-			float gate_limit = sqrtf((sq(gate_sigma) * aid_src_status.innovation_variance));
-			aid_src_status.innovation = math::constrain(aid_src_status.innovation, -gate_limit, gate_limit);
-
-			// also reset the yaw gyro variance to converge faster and avoid
-			// being stuck on a previous bad estimate
-			resetZDeltaAngBiasCov();
+		// set the magnetometer unhealthy if the test fails
+		if (aid_src_status.innovation_rejected) {
+			_innov_check_fail_status.flags.reject_yaw = true;
+
+			// if we are in air we don't want to fuse the measurement
```

Comment:
I think we could even consider removing this since we now have the zero gyro fusion when on ground and at rest